File: Irrgarten/src/irrgarten/Dice.java

package irrgarten;

import java.util.Random;

public class Dice {
    private static final int MAX_USES = 5; //Usos máximos de armas y escudos
    private static final float MAX_INTELLIGENCE = 10.0f; //Valor máximo de inteligencia para monstruso y jugadores
    private static final float MAX_STRENGTH = 10.0f; //Valor máximo de fuerza para monstruos y jugadores
    private static final float RESURRECT_PROB = 0.3f; //Probabilidad de que un jugador sea resucitado en cada turno
    private static final int WEAPONS_REWARD = 2; //Número máximo de armas recibidas al ganar un combate
    private static final int SHIELDS_REWARD = 3; //Número máximo de escudos recibidos al ganar un combate
    private static final int HEALTH_REWARD = 5; //Saludo máxima recibida al ganar un combate
    private static final float MAX_ATTACK = 3.0f; //Potencia máxima de armas
    private static final float MAX_SHIELD = 2.0f; //Protección máxima de escudos
    
    private static final Random generator = new Random();
    
    public int randomPos (int max){
        
        return generator.nextInt(max);
    }
    
    public int whoStarts (int nplayers){
        
        return generator.nextInt(nplayers);
    }
    
    public float randomIntelligence(){
        return generator.nextFloat(MAX_INTELLIGENCE);
    }
    
    public float randomStrength(){
        return generator.nextFloat(MAX_STRENGTH);
    }
    
    public boolean resurrectPlayer(){
        boolean resucita = false;
        if (generator.nextFloat(1.0f) < 0.3) {
            resucita = true;
        }
        return resucita;
    }
    
    public int weaponsReward() {
        return generator.nextInt(WEAPONS_REWARD+1);
    }
    
    public int shieldsReward(){
        return generator.nextInt(SHIELDS_REWARD+1);
    }
    
    public int healthReward(){
        return generator.nextInt(HEALTH_REWARD+1);
    }
    
    public float weaponPower(){
        return generator.nextFloat(MAX_ATTACK);
    }
    
    public float shieldPower(){
        return generator.nextFloat(MAX_SHIELD);
    }
    
    public int usesLeft(){
        return generator.nextInt(MAX_USES+1);
    }
    
    public float intensity(float competence){
        return generator.nextFloat(competence);
    }
    
    public boolean discardElement(int usesLeft){
        boolean descartado = false;
        float probabilidad = 1.0f - ((float) usesLeft/(float) MAX_USES);
        if (generator.nextFloat(1.0f) < probabilidad){
            descartado = true;
        }
        
        return descartado;
    }
}

File: Irrgarten_Ruby/Dice.rb
#encoding:utf-8

module Irrgarten 
	class Dice
	
		@@MAX_USES = 5 #Usos máximos de armas y escudos
		@@MAX_INTELLIGENCE = 10.0 #Valor máximo de inteligencia para monstruso y jugadores
		@@MAX_STRENGTH = 10.0 #Probabilidad de que un jugador sea resucitado en cada turno
		@@RESURRECT_PROB = 0.3 #Número máximo de armas recibidas al ganar un combate
		@@WEAPONS_REWARD = 2 #Número máximo de escudos recibidos al ganar un combate
		@@SHIELDS_REWARD = 3 #Número máximo de escudos recibidos al ganar un combate
		@@HEALTH_REWARD = 5 #Saludo máxima recibida al ganar un combate
		@@MAX_ATTACK = 3.0 #Potencia máxima de armas
		@@MAX_SHIELD = 2.0	#Protección máxima de escudos
			
		@@generator = Random.new
		
		def self.random_pos ( max ) #int
			@@generator.rand(0...max) # ... excluido .. incluido
		 
		end
		
		
		def who_starts ( nplayers ) #int
			@@generator.rand(0...nplayers)
		
		end
		
		
		def self.random_intelligence()
			@@generator.rand(0...@@MAX_INTELLIGENCE)
		end
		
		
		def self.random_strength()
			@@generator.rand(0...@@MAX_STRENGHT)
		end
		
		
		def self.resurrect_player()
			if (@@generador.rand < 0.3)
				return true
			else
				return false
		end
		
		
		def self.weapons_reward()
			@@generator.rand(0..@@WEAPONS_REWARD)
		end
		
		
		def self.shields_reward()
			@@generator.rand(0..@@SHIELDS_REWARD)
		end
		
		
		def self.health_reward()
			@@generator.rand(0..@@HEALTH_REWARD)
		end
		
		
		def self.weapon_power()
			@@generator.rand(0...@@MAX_ATTACK)
		end
		
		
		def self.shield_power()
			@@generator.rand(0...@@MAX_SHIELD)
		end
		
		
		def self.uses_left()
			@@generator.rand(0..@@MAX_USES)
		end
		
		
		def selfintensity( competence )
			@@generator.rand(0...competence)
		end
		
		
		def self.discard_element( uses_left )
			descartado = false
			probabilidad = 1.0 -  (uses_left.to_f/@@MAX_USES.to_f)
			if @@generator.rand(0.0..1.0) < probabilidad
			    descartado = true
			end
			
			return descartado
		end
	end	
end
